### PR TITLE
Fix and improve Polargraph

### PR DESCRIPTION
**Pull request by i-make-robots**
### Description

Fixes for 2.1.x to work with polargraph.  a few sanity checks were too broad and motion stuff was wonky.

### Requirements

Tested with RUMBA board.  Should also work with RAMPS.

### Benefits

Ensures POLARGRAPH support stays current.

### Configurations

Taken directly from previous PR on `2.1.x`
[Marlin.zip](https://github.com/MarlinFirmware/Marlin/files/9739907/Marlin.zip)

### CODE REVIEW

File: Marlin/src/inc/SanityCheck.h [hunk 1]
```diff
@@ -822,15 +822,17 @@ static_assert(Y_MAX_LENGTH >= Y_BED_SIZE, "Movement bounds (Y_MIN_POS, Y_MAX_POS
  * Granular software endstops (Marlin >= 1.1.7)
  */
 #if ENABLED(MIN_SOFTWARE_ENDSTOPS) && DISABLED(MIN_SOFTWARE_ENDSTOP_Z)
-  #if IS_KINEMATIC
+  #if ENABLED(POLARGRAPH)
+  #elif IS_KINEMATIC
```
Comment (thisiskeithb) [inline Marlin/src/inc/SanityCheck.h]: I think you're missing an `#error` here. There's nothing in this `#if`, so it's simply skipped.

File: Marlin/src/inc/SanityCheck.h [hunk 2]
```diff
@@ -822,15 +822,17 @@ static_assert(Y_MAX_LENGTH >= Y_BED_SIZE, "Movement bounds (Y_MIN_POS, Y_MAX_POS
...
     #error "MIN_SOFTWARE_ENDSTOPS on DELTA/SCARA also requires MIN_SOFTWARE_ENDSTOP_Z."
   #elif NONE(MIN_SOFTWARE_ENDSTOP_X, MIN_SOFTWARE_ENDSTOP_Y)
     #error "MIN_SOFTWARE_ENDSTOPS requires at least one of the MIN_SOFTWARE_ENDSTOP_[XYZ] options."
   #endif
 #endif
 
 #if ENABLED(MAX_SOFTWARE_ENDSTOPS) && DISABLED(MAX_SOFTWARE_ENDSTOP_Z)
-  #if IS_KINEMATIC
+  #if ENABLED(POLARGRAPH)
+  #elif IS_KINEMATIC
```
Comment (thisiskeithb) [inline Marlin/src/inc/SanityCheck.h]: Ditto here.

File: Marlin/src/inc/SanityCheck.h (hunk 1 above)
Comment (i-make-robots) [reply Marlin/src/inc/SanityCheck.h]: there should be nothing.  the IS_KINEMATIC error is intended for deltas, not polargraph.

Comment (thisiskeithb) [reply Marlin/src/inc/SanityCheck.h]: Then the check should be added to the parent `#if`:

`#if ENABLED(MIN_SOFTWARE_ENDSTOPS) && NONE(MIN_SOFTWARE_ENDSTOP_Z, POLARGRAPH)`

**Comment by i-make-robots** [comment]
M0_M1.cpp gives me a problem.
```c
- TERN_(HOST_PROMPT_SUPPORT, hostui.prompt_do(PROMPT_USER_CONTINUE, parser.codenum ? F("M1 Stop") : F("M0 Stop"), FPSTR(CONTINUE_STR)));
+  #if ENABLED(HOST_PROMPT_SUPPORT)
+    hostui.prompt_do(PROMPT_USER_CONTINUE, 
+    parser.string_arg ? parser.string_arg : (parser.codenum ? F("M1 Stop") : F("M0 Stop")), FPSTR(CONTINUE_STR));
+  #endif```
problem is prompt_do wants an `FSTR_P`, but parser.string_arg is a `static char *`.  I asked in Discord and no one knew how to convert the one to the other.

**Comment by thinkyhead** [comment]
We still cannot push patches to your PRs. At some point in the future, try deleting your fork and creating a new one just using the name "Marlin", and maybe it will solve this issue. In the meantime, it would be most convenient if you added me as a collaborator on your fork so that I am able to push changes instead of submitting PRs to your branch.

File: Marlin/Configuration.h [hunk 3]
```diff
@@ -87,7 +87,7 @@
 
 // Choose the name from boards.h that matches your setup
 #ifndef MOTHERBOARD
-  #define MOTHERBOARD BOARD_RAMPS_14_EFB
+  #define MOTHERBOARD BOARD_RUMBA
```
Comment (thinkyhead) [inline Marlin/Configuration.h]: Obviously we cannot push this PR with your configurations in it. The idea with PRs is to set them up exactly as they should be published for everybody.

File: Marlin/src/module/polargraph.cpp [hunk 4]
```diff
@@ -44,7 +44,7 @@ xy_pos_t draw_area_min = { X_MIN_POS, Y_MIN_POS },
 
 xy_float_t draw_area_size = { X_MAX_POS - X_MIN_POS, Y_MAX_POS - Y_MIN_POS };
 
-float polargraph_max_belt_len = HYPOT(draw_area_size.x, draw_area_size.y);
+float polargraph_max_belt_len = POLARGRAPH_MAX_BELT_LEN;
```
Comment (thinkyhead) [inline Marlin/src/module/polargraph.cpp]: None of these configurable values is being saved to EEPROM, so code to load/save/reset should be added to `settings.cpp`.

**Comment by thinkyhead** [comment]
Here are the patches to get this PR closer to done… MarginallyClever/Marlin-polargraph#5

File: Marlin/src/module/motion.cpp [hunk 5]
```diff
@@ -562,7 +561,8 @@ void do_blocking_move_to(NUM_AXIS_ARGS(const float), const_feedRate_t fr_mm_s/*=
     const feedRate_t w_feedrate = fr_mm_s ?: homing_feedrate(W_AXIS)
   );
 
-  #if IS_KINEMATIC
+  #if IS_KINEMATIC && DISABLED(POLARGRAPH)
+    // kinematic machines are expected to home to a point 1.5x their range? never reachable.
```
Comment (thinkyhead) [inline Marlin/src/module/motion.cpp:565]: Is this referring to a Delta homing towards the top of the machine? If so, note that `position_is_reachable` only checks the x, y position to see if it's within the bed area, whether square or round-ish.

**Comment by thinkyhead** [comment]
> M0_M1.cpp gives me a problem.

If you want to send an arbitrary C-string stored in SRAM you'll need to create a new implementation of `HostUI::prompt_do` that accepts a C-string. The aim with `HostUI` is that all messages are standardized, so that capability has not been added.

**Comment by thinkyhead** [comment]
The variable `draw_area_size` is never used, so I went ahead and removed it.

**Pull request merged**